// handybars/src/index.js

import lexer from './lexer';
import { isString, wtf, isObject, set, makeContext, safe, safeJoin } from './utils';
import {
	Text,
	Block,
	Invocation,
	Collection,
	Identifier,
	Literal,
	CompoundIdentifier,
} from './taxonomy';
import DefaultHelpers from './helpers';

export function parse (input) {
	if (isString(input)) return lexer(input);
	if ((input instanceof Block)) return input;
	wtf('Parse can only receive strings or a Handybars AST tree');
}

export function evaluate (input, ctx) {
	if (isString(input)) input = lexer(input);
	else if (!(input instanceof Block)) {
		wtf('Evaluate can only receive strings or a Handybars AST tree');
	}

	return input.evaluate(ctx, DefaultHelpers).value;
}

export {
	parse as partial,
	Text,
	Block,
	Invocation,
	Collection,
	Identifier,
	Literal,
	CompoundIdentifier,
	makeContext,
	safe,
	safeJoin,
};

export default function Handybars (template, world = {}) {
	const ast = parse(template);
	const env = { ...DefaultHelpers, ...world };

	function execute (ctx) {
		const frame = makeContext(ctx, env);
		return ast.evaluate(ctx, frame).value;
	}

	execute.set = (...args) => {
		if (isObject(args[0])) {
			Object.assign(env, ...args);
		} else {
			set(env, ...args);
		}
		return execute;
	};

	execute.setPartial = (name, subtemplate) => set(env, name, lexer(subtemplate));

	return execute;
}
